### PR TITLE
Support Artifact Bundle

### PR DESCRIPTION
**Pull request by freddi-kit**
# Motivation
[SE-0305](https://github.com/apple/swift-evolution/blob/main/proposals/0305-swiftpm-binary-target-improvements.md) is introduced and it is now important for build tools. 

Artifact Bundle helps to support Artifact bundle to not take time to making build binary. 

I developed new OSS tool named ArtifactBundleGen to create it easily.

# Changes
- Add Artifact Bundle creation flow

# Test
Run `make archive`

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,10 @@
 
 ## Next Version
 
+### Added
+
+- Support Artifact Bundle # @freddi-kit
```
Comment (freddi-kit) [inline CHANGELOG.md:7]:
```suggestion
- Support Artifact Bundle #1388 @freddi-kit
```

File: Makefile [hunk 2]
```diff
@@ -48,3 +48,7 @@ brew:
 
 archive: build
 	./scripts/archive.sh "$(EXECUTABLE_PATH)"
+	swift package plugin --allow-writing-to-package-directory generate-artifact-bundle \
+		--package-version $(VERSION) \
+		--executable-name xcodegen \
```
Comment (yonaskolb) [inline Makefile:53]:
```suggestion
		--executable-name $(EXECUTABLE_NAME) \
```

**Review by yonaskolb — COMMENTED** [review]
Thanks @freddi-kit. 

What is the story around the `SettingPresets` assets? I don't think these are being bundled with the artifact. I assume we need to move these to package resources first.

As an aside, would love the github release process to be automated at some point.

**Comment by freddi-kit** [comment]
> What is the story around the SettingPresets assets? I don't think these are being bundled with the artifact. I assume we need to move these to package resources first.

I support it in this commit, could you help checking it? 🙇 

> As an aside, would love the github release process to be automated at some point.

I have idea to do it, let me do it in other PR 👍

**Comment by freddi-kit** [comment]
i noticed i need to sereval things for support resource please wait a moment

**Comment by freddi-kit** [comment]
@yonaskolb 
https://github.com/yonaskolb/XcodeGen/pull/1388/commits/512ec959f2ace2d96a77d77d394562725d9ea16b
Now SettingPresets can be loaded from artifact bundle, please check it

File: Package.swift [hunk 3]
```diff
@@ -2,6 +2,13 @@
 
 import PackageDescription
 
+let macOSOnlyDependency: [Package.Dependency]
+#if !os(Linux)
```
Comment (yonaskolb) [inline Package.swift:6]: Any reason not to do a mac check here instead to be more explicit?

File: Sources/XcodeGenKit/SettingPresets [hunk 4]
```diff
@@ -0,0 +1 @@
+../../SettingPresets/
```
Comment (yonaskolb) [inline Sources/XcodeGenKit/SettingPresets:1]: What does this file do?

Comment (freddi-kit) [reply Sources/XcodeGenKit/SettingPresets:1]: This is symbolic link for SeetingPresets in root to detect it as bundle file for XcodeGenKit. Package.swift cannot find it if we don't have it in root of target source file here https://github.com/yonaskolb/XcodeGen/pull/1388/files#diff-f913940c58e8744a2af1c68b909bb6383e49007e6c5a12fb03104a9006ae677eR52

We also cannot set it as .copy("../../SettingPresets")

File: Package.swift (hunk 3 above)
Comment (freddi-kit) [reply Package.swift:6]: In CI machine, we cannot build my tools, I'm checking the reason

**Comment by freddi-kit** [comment]
@yonaskolb Linux issue is solved, please check again

File: Makefile [hunk 5]
```diff
@@ -48,3 +48,7 @@ brew:
 
 archive: build
 	./scripts/archive.sh "$(EXECUTABLE_PATH)"
+	swift package plugin --allow-writing-to-package-directory generate-artifact-bundle \
+		--package-version $(VERSION) \
+		--executable-name $(EXECUTABLE_NAME) \
+		--build-config release \
```
Comment (freddi-kit) [inline Makefile:54]:
```suggestion
		--build-config release
```

File: Package.swift (hunk 3 above)
Comment (freddi-kit) [reply Package.swift]: fixed

**Comment by roman-aliyev** [comment]
@yonaskolb @freddi-kit We have recently solved the same task. See mac-cain13/R.swift#838 . May be this will help you.

**Comment by freddi-kit** [comment]
@yonaskolb do you have any concern?

File: Sources/XcodeGenKit/SettingsBuilder.swift [hunk 6]
```diff
@@ -202,6 +202,9 @@ extension SettingsPresetFile {
                 symlink.parent() + relativePath,
             ] + possibleSettingsPaths
         }
+        if let moduleResourcePath = Bundle.module.path(forResource: "SettingPresets", ofType: nil) {
+            possibleSettingsPaths.append(Path(moduleResourcePath) + "\(path).yml")
+        }
```
Comment (freddi-kit) [inline Sources/XcodeGenKit/SettingsBuilder.swift:207]: @roman-aliyev If @yonaskolb agree with your way, please open PR and do not forget to add this code and copy script SettingPresets to bundle.

Comment (freddi-kit) [reply Sources/XcodeGenKit/SettingsBuilder.swift:207]: Also, please add https://github.com/yonaskolb/XcodeGen/pull/1388/files#diff-f913940c58e8744a2af1c68b909bb6383e49007e6c5a12fb03104a9006ae677eR45-R46 and make symbolic link as https://github.com/yonaskolb/XcodeGen/pull/1388/files#diff-9f320f241fc06216ac0b9fed9a549298a11c8f203c299350159cab6946344b14

**Comment by mat1th** [comment]
Is there any update on this PR? I would like to use `XcodeGen` from a swiftPackage.

**Comment by freddi-kit** [comment]
@yonaskolb could you help checking it again? 🙇

**Review by yonaskolb — APPROVED** [review]
Looks good to me, thanks @freddi-kit 🙏

**Comment by kkk669** [comment]
I created a sample project. I hope this will be helpful for someone.

https://github.com/kkk669/xcodegen-artifactbundle-demo/

**Pull request merged**